### ISSUE TITLE
Changes no longer detected in files that have been rotated and ignore_older kicks in before the next change

### ISSUE THREAD

**Issue by pghalliday** [elastic/beats#596]
Raised this yesterday here: https://github.com/elastic/filebeat/issues/320
Today I realised that the project has moved.

---

Using the following configuration and Filebeat v1.0.1 on Ubuntu 14.04

```
filebeat:
  prospectors:
    - paths:
        - "/var/log/test/*.log"
      ignore_older: 30s
      scan_frequency: 5s

  registry_file: /var/lib/filebeat/registry

output:
  logstash:
    hosts: ["logstash.myserver.com:5044"]

shipper:

logging:
  level: debug
  to_files: true
  files:
    path: /var/log/filebeat
    name: filebeat.log
    rotateeverybytes: 10485760 # = 10MB
```

Changes to files are not detected when rotating files then waiting 30s before adding another line. ie:

1. `mkdir -p /var/log/test`
1. Start Filebeat with no log files in /var/log/test/
1. `cd /var/log/test`
1. `echo "hello" >> hello.log`
1. Line is picked up and published
1. Wait 30s for ignore_older to kick in
1. `echo "hello" >> hello.log`
1. Line is picked up and published
1. `mv hello.log hello.log.1 && echo "hello" >> hello.log`
1. Line is picked up and published
1. `mv hello.log hello.log.1 && touch hello.log`
1. Wait 30s for ignore_older to kick in
1. `echo "hello" >> hello.log`
1. Line is **NOT** picked up and published

There are no additional errors in the debug, just lots of this:

```
2015-12-28T14:50:51Z DBG  Start next scan
2015-12-28T14:50:51Z DBG  scan path /var/log/test/*.log
2015-12-28T14:50:51Z DBG  Check file for harvesting: /var/log/test/hello.log
2015-12-28T14:50:51Z DBG  Update existing file for harvesting: /var/log/test/hello.log
2015-12-28T14:50:51Z DBG  Not harvesting, file didn't change: /var/log/test/hello.log
2015-12-28T14:50:55Z DBG  Flushing spooler because of timemout. Events flushed: 0
```

Filebeat seems to continue and in more realistic scenarios, other prospectors continue working and publishing logs.

**Comment by pghalliday** [comment]
I've been trying to create a test case for this in `filebeat/tests/system/test_prospector.py` and this is what I have:

```python
    def test_rotating_ignore_older_low_write_rate(self):
        self.render_config_template(
            path=os.path.abspath(self.working_dir) + "/log/*",
            ignoreOlder="1s",
            scan_frequency="0.1s",
        )

        os.mkdir(self.working_dir + "/log/")
        testfile = self.working_dir + "/log/test.log"

        proc = self.start_filebeat(debug_selectors=['*'])
        time.sleep(1)

        lines = 0

        lines += 1
        with open(testfile, 'a') as file:
            file.write("Line {}\n".format(lines))
        # wait for log to be read
        time.sleep(0.2)
        # log rotate
        os.rename(testfile, testfile + ".1")
        os.mknod(testfile)
        # wait for ignore older
        time.sleep(2)
        lines += 1
        with open(testfile, 'a') as file:
            file.write("Line {}\n".format(lines))

        self.wait_until(
            # allow for events to be send multiple times due to log rotation
            lambda: self.output_count(lambda x: x >= lines),
            max_timeout=5)

        proc.kill_and_wait()
```

This fails on the `v1.0.1` tag and the `1.0.1` branch. However it succeeds on the `master` and `1.1` branches. So it looks like it has been fixed, not sure which commit though. I'll leave the test here in case you want to add it for regression testing... or I can submit a pull request if you want.

Anyway, when can we expect a new release with the fix? :)

The issue itself seems quite severe to me

**Comment by ruflin** [comment]
@pghalliday Thanks a lot for the detailed report and also the test. Looking at https://github.com/elastic/filebeat/issues/181 I'm not sure why the above happens under 1.0.1 and not 1.1. The first thing I was thinking of is that the scan_frequency is larger then ignore_older, but that is not the case in your test above.

Could you open a PR with the test? Like this I can also test it automated with the 1.0 branch and probably figure out what the issue is / was.

**Comment by pghalliday** [comment]
No problem, PR #601 submitted. While I was debugging this I did notice that the `harvester.FileStat` object did not have a `Finished()` method on v1.0.1. This was going to be my next line of investigation as I figured it must have been a change in the `ignore_older` implementation as used [here](https://github.com/elastic/beats/blob/0d147f729680f3205e2afd203239e86fc4b07b39/filebeat/crawler/prospector.go#L416)

The check was different in the v1.0.1 tag, [here](https://github.com/elastic/beats/blob/v1.0.1/filebeat/crawler/prospector.go#L374)

The change seems to be from [this commit](https://github.com/elastic/beats/commit/71e5b2265b5d900c79ee97ffdc8779e445f5a6b7)

Could be a red herring though, I decided to stop as it was fixed eitherway (I don't really know go never mind this code base!) - i figured the test might help keep it fixed :)

**Comment by ruflin** [comment]
Yes, after 1.0.0 we did quite a bit of refactoring and reorganisation of the code. Now with the multiline feature some more changes came in which actually made the code simpler. But because of the bigger changes we decided to ship it with 1.1. Most of the changes kept the logic the same but made the code simpler and more readable. An other thing we target is race conditions. With very short cycles or bad timing it is/was possible to create race conditions. We eliminate quite a few cases so far but not all, for example https://github.com/elastic/filebeat/issues/300 The reason I'm not sure it is a race condition is that it can be reproduced so reliably. 

Having the test is definitively the best case to make sure it stays fixed. What I want to figure out is what the issue in 1.0.1 is exactly and if we should ship a fix with 1.0.2. Version 1.1. should be out in January.

**Comment by pghalliday** [comment]
I don't think it's a race condition either - I originally encountered the issue with quiet nginx logs being rotated daily with a 24 hour `ignore_older` setting. The log rotation and ignore_older point will have been hours apart and it was still 100%. One day I would be hitting the server for a few hours then leave it for days, the next time I hit the server the logs were no longer being harvested and the harvester didn't restart. I'm pretty sure the important factor is that a new harvester is started for a known filename (because of rotation) that then gets ignored before any changes are detected

**Comment by pghalliday** [comment]
I've added a comment to the pull request - I now suspect a problem in the rename functionality in v1.0.1

**Comment by ruflin** [comment]
Closing as this was fixed in 1.1
